3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/UnbounceAdapter.scala

```diff
@@ -85,39 +88,44 @@ object UnbounceAdapter extends Adapter {
+      case (_, None)                          => s"Request body provided but content type empty, expected ${ContentType} for ${VendorName}".failNel
+      case (_, Some(ct)) if ct != ContentType => s"Content type of ${ct} provided, expected ${ContentType} for ${VendorName}".failNel
+      case (Some(body),_)                     => {
         if (body.isEmpty) {
           s"${VendorName} event body is empty: nothing to process".failNel
         } else {
+          val qsParams = toMap(payload.querystring)
           try {
             val bodyMap = toMap(URLEncodedUtils.parse(URI.create("http://localhost/?" + body), "UTF-8").toList)
-            (payloadBodyToEvent(bodyMap), payloadBodyToContext(bodyMap)) match {
-              case (Failure(err), _) => err.failNel
-              case (_, Failure(err)) => err.failNel
-              case (Success(event), Success(context)) => {
-                SchemaKey.parse(schemaUri) match {
-                  case Failure(procMsg) => procMsg.getMessage.failNel
-                  case Success(_) => {
-                    NonEmptyList(RawEvent(
-                      api         = payload.api,
-                      parameters  = toUnstructEventParams(TrackerVersion, params, schemaUri, event, context, "srv"),
-                      contentType = payload.contentType,
-                      source      = payload.source,
-                      context     = payload.context
-                    )).success
-                  }
-                }
+            val json = compact(render(bodyMap))
+            val event = parse(json)
+            lookupSchema(Some("event_context"), VendorName, ContextSchema) match {
+              case Failure(str)  => str.fail
+              case Success(schema) => {
+               val eventWithData = bodyMap.get("data.json")
+                 .flatMap(s => Try(parse(s)).toOption)
+                 .flatMap(dj => Some(event.replace(List("data.json"), dj)))
+                 .getOrElse(event)
+                NonEmptyList(RawEvent(
+                  api          = payload.api,
+                  parameters   = toUnstructEventParams(TrackerVersion, qsParams, schema, eventWithData, "srv"),
+                  contentType  = payload.contentType,
+                  source       = payload.source,
+                  context      = payload.context
+                )).success
               }
             }
-          } catch { 
+
+          } catch {
+            case e: JsonParseException => {
+              val exception = JU.stripInstanceEtc(e.toString).orNull
+              s"${VendorName} event string failed to parse into JSON: [${exception}]".failNel
+            }
             case e: Exception => {
               val exception = JU.stripInstanceEtc(e.toString).orNull
-              s"${VendorName} could not parse body: [${exception}]".failNel
+              s"${VendorName} incorrect event string : [${exception}]".failNel
             }
```

Comment:
same here, let's remove those curlies

Reply:
Don't you think that this will make the resulting long line less readable? Also what do you think about cathing Exception? Is it not too general?